tests/test_arrow_dataset.py

```diff
@@ -3629,6 +3630,39 @@ def test_dataset_from_generator_features(features, data_generator, tmp_path):
     _check_generator_dataset(dataset, expected_features)
 
 
+def test_from_spark():
+    spark = pyspark.sql.SparkSession.builder.appName("pyspark").getOrCreate()
+    data = [
+        ("0", 0, 0.0),
+        ("1", 1, 1.0),
+        ("2", 2, 2.0),
+        ("3", 3, 3.0),
+    ]
+    df = spark.createDataFrame(data, "col_1: string, col_2: int, col_3: float")
+    dataset = Dataset.from_spark(df)
```

Comment:
Could you add a test to make sure that calling `from_spark` twice in a row with two different DataFrames results in two different Dataset objects ?

The `DatasetBuilder` uses a `cache_dir` with subdirectories defined after the name of the builder ("spark") and the arguments of the dataset to generate. Here since the `**kwargs` passed to `DatasetBuilder.__init__` won't change between two `from_spark`, it may reuse the same cache and return the same dataset.

If this is the case we can add a `df_id` parameter to the `SparkConfig`, which can be set to to be the python `id` of the Spark DataFrame for example (or any identifier of the Spark DataFrame).

Reply:
I modified the DatasetBuilder to cache based on the dataframe's semanticHash and added a test for it. This isn't perfect; e.g if a user generates a dataset from query "SELECT * FROM T", adds rows to the table, and then regenerates the dataset, the semantic hash for the dataframe will be the same, so the user will get the stale data. However, they can pass the force_download=True option to from_spark to force redownloading. Wdyt?

Reply:
Looks good ! (I was off the last few days, sorry for the late reply)

Reply:
Unfortunately no, the same df will generate a different a semantic hash in different Spark sessions.